src/Moq/Expectation.cs

```diff
@@ -0,0 +1,45 @@
+// Copyright (c) 2007, Clarius Consulting, Manas Technology Solutions, InSTEDD, and Contributors.
+// All rights reserved. Licensed under the BSD 3-Clause License; see License.txt.
+
+using System;
+using System.Linq.Expressions;
+
+using Moq.Async;
+
+namespace Moq
+{
+	/// <summary>
+	///   Represents a set (or the "shape") of invocations
+	///   against which concrete <see cref="Invocation"/>s can be matched.
+	/// </summary>
+	internal abstract class Expectation : IEquatable<Expectation>
+	{
+		public abstract LambdaExpression Expression { get; }
+
+		public virtual bool HasResultExpression(out IAwaitableFactory awaitableFactory)
+		{
+			awaitableFactory = null;
+			return false;
+		}
+
+		public override bool Equals(object obj)
+		{
+			return obj is Expectation other && this.Equals(other);
+		}
+
+		public abstract bool Equals(Expectation other);
+
+		public abstract override int GetHashCode();
+
+		public abstract bool IsMatch(Invocation invocation);
+
+		public virtual void SetupEvaluatedSuccessfully(Invocation invocation)
+		{
+		}
```

Comment:
This also doesn't fit in very well here.

I'm not yet sure how to get rid of it. A few possible options off the top off my head:

 * Let `Expectation` expose its argument matchers through a property. When a setup is successfully matched, `Setup` could then skip over `Expectation.SetupEvaluatedSuccessfully` and call the matchers' `SetupEvaluatedSuccessfully` methods directly.

 * Add an event `Setup.Matched`, to which argument matchers may optionally subscribe their `SetupEvaluatedSuccessfully` methods.

The first option seems like the more plausible in the short term. It doesn't make much sense to add a `Matched` event just to get rid of this method here, however that solution may start to make more sense if the event is going to get added to also support other features (such as better support for sequences).